src/service/video-manager-impl.js

```diff
@@ -0,0 +1,310 @@
+
+    // TODO(aghassemi, #4780): Create a new IntersectionObserver service.
+    if (!this.scrollListenerInstalled_) {
+      const scrollListener = () => {
+        for (let i = 0; i < this.entries_.length; i++) {
+          this.entries_[i].updateVisibility();
+        }
+      };
+      const viewport = viewportFor(this.win_);
+      viewport.onScroll(scrollListener);
+      viewport.onChanged(scrollListener);
+      this.scrollListenerInstalled_ = true;
+    }
+  }
+}
+
+/**
+ * VideoEntry represents an entry in the VideoManager's list.
+ */
+class VideoEntry {
+  /**
+   * @param {!Window} win
+   * @param {!../video-interface.VideoInterface} video
+   */
+  constructor(win, video) {
+
+    /** @package @const {!../video-interface.VideoInterface} */
+    this.video = video;
+
+    /** @private {boolean} */
+    this.loaded_ = false;
+
+    /** @private {boolean} */
+    this.isVisible_ = false;
+
+    /** @const @private {!../../src/service/vsync-impl.Vsync} */
+    this.vsync_ = vsyncFor(win);
+
+    /** @private {boolean} */
+    this.canAutoplay_ = video.hasAutoplay() &&
+        platformSupportsAutoplay(platformFor(win));
+
+    const element = dev().assert(video.element);
+
+    listenOncePromise(element, VideoEvents.BUILT)
+      .then(() => this.videoBuilt_());
+
+    listenOncePromise(element, VideoEvents.CAN_PLAY)
+      .then(() => this.videoLoaded_());
+  }
+
+  /**
+   * Called when the video element is built.
+   * @private
+   */
+  videoBuilt_() {
+    if (this.canAutoplay_) {
+      this.autoplayVideoBuilt_();
+    }
+  }
+
+  /**
+   * Called when the video is loaded and can play.
+   * @private
+   */
+  videoLoaded_() {
+    this.loaded_ = true;
+    if (this.isVisible_) {
+      // Handles the case when the video becomes visible before loading
+      this.loadedVideoVisibilityChanged_();
+    }
+  }
+
+  /**
+   * Called when visibility of a video changes.
+   * @private
+   */
+  videoVisibilityChanged_() {
+    if (this.loaded_) {
+      this.loadedVideoVisibilityChanged_();
+    }
+  }
+
+  /**
+   * Only called when visibility of a loaded video changes.
+   * @private
+   */
+  loadedVideoVisibilityChanged_() {
+    if (this.canAutoplay_) {
+      this.autoplayLoadedVideoVisibilityChanged_();
+    }
+  }
+
+  /**
+   * Whether there is a need to monitor visibility of this video.
+   * @return {boolean}
+   * @package
+   */
+  needsVisibilityObserver() {
+    return this.canAutoplay_;
+  }
+
+  /* Autoplay Behaviour */
+
+  /**
+   * Called when an autoplay video is built.
+   * @private
+   */
+  autoplayVideoBuilt_() {
+    this.video.mute();
+
+    // If autoplay video has controls, hide them and only show them on
+    // user-ineraction.
+    if (this.video.hasControls()) {
+      this.video.hideControls();
+
+      // TODO(aghassemi): This won't work for iframes, needs a transparent shim
+      listenOnce(this.video.element, 'click', () => {
+        this.video.showControls();
+        this.video.unmute();
```

Comment:
deprecated `muted` attribute.
We may not end up unmuting here depending on what comes out of UX review tomorrow however.